src/imap/imap-client.c

```diff
@@ -552,8 +586,84 @@ void client_add_capability(struct client *client, const char *capability)
 		/* explicit capability - don't change it */
 		return;
 	}
-	str_append_c(client->capability_string, ' ');
-	str_append(client->capability_string, capability);
+
+    /* add it to our capability list as CAP_ALWAYS */
+    imap_capability_list_add(client->capability_list,
+            capability, CAP_ALWAYS);
+}
+
+const char *client_get_capability(struct client *client)
+{
+	string_t *cap_str = t_str_new(256);
+
+	/* iterate capability list and append each capability if conditions met */
+	const struct imap_capability *capp;
+	array_foreach(&client->capability_list->imap_capabilities, capp) {
+		/* local flags variable will get filled with all the flags
+		   that matched the current conditions -- then at the end
+		   if FLAG_REQUIRE_ALL is present the local flags variable
+		   is compared against the command flags to check for a
+		   perfect match */
```

Comment:
Too much code duplication here. Maybe instead imap_capability_list_get_capabilities() where the caller gives the current CAP_* as parameter? So something like:

```
flags = CAP_POSTAUTH;
if (client->starttls)
  flags |= CAP_TLSACTIVE;
else
  flags |= CAP_TLSINACTIVE;
...
imap_capability_list_get_capabilities(cap_str, flags);

Reply:
Yes, I like this approach much better.

If the imap_capability_list_get_capabilities() function does not distinguish between the meaning of each IMAP_CAP_VISIBILITY_ flag then it could be done in a general form with a for-loop like this:

        uint32_t flag_accumulator = 0;
        uint32_t n = 0;
        if (capp->flags) {
            /* iterate all known IMAP_CAP_VISIBILITY_ flags */
            for (n = 0; n < 9; ++n) {
                if ((capp->flags & BIT(n)) != 0) {
                    flag_accumulator |= n;
                }
            }
            bool add = FALSE;
            /*  now check if CAP_REQUIRE_ALL is present */
            if ((capp->flags & CAP_REQUIRE_ALL) != 0) {
                /*  did we match all? */
                if ((capp->flags & ~CAP_REQUIRE_ALL) == flags) {
                    add = TRUE;
                }
            /*  no CAP_REQUIRE_ALL - did we match any flags? */
            } else if (flags > 0) {
                add = TRUE;
            }
            /*  should we add this entry? */
            if (add) {
                if (*str_c(cap_str) != '\0') {
                    str_append_c(cap_str, ' ');
                }
                str_append(cap_str, capp->command);
            }
        }

As opposed to checking the individual flags and the whether the conditions they require are met:

        /*  is starttls active? */
        if ((capp->flags & CAP_TLSACTIVE) != 0 &&
                        client->starttls) {
                flags |= CAP_TLSACTIVE;
        }

Do you have any reservations about the generalized for-loop approach, for example an issue I have with the for-loop approach is that the 9 is hardcoded into the for-loop being the highest bit-level that IMAP_CAP_VISIBILITY_ flags reach (0 - 8). Do you know if there is a better way to do this without hardcoding the 9?

I often use the counting property of enums to produce a _count entry:

    enum val_enum {
        val_invalid = -1, 

        val_0,
        val_1,
        val_2,

        val_count // always reflects the total number of val_ entries
    };

This allows for checks like:

    if (val <= val_invalid || val >= val_count) {
        // val is outside valid bounds
    }

And iteration in the form of:

    enum val_enum val = val_invalid;
    for (val = val0; val < val_count; ++val) {
        /* iterates all enumerated val_ entries */
    }

But the above approach would need some adjustments to be applied in the context of bitflags and enumerated bitflags.

Of course I don't have to condense into a for-loop, I could just leave the imap_capability_list_get_capabilities() function to manually check for each IMAP_CAP_VISIBILITY_ flag and manually OR the flag into the flag_accumulator the way it was being done previously.

Any suggestions at all would be appreciated!